Week1/homework/app.js

```diff
@@ -103,60 +103,45 @@
   };
 
-  function createHtml() {
-    const mainHeadings = document.createElement('h1');
-    document.body.appendChild(mainHeadings);
-    const mainHeadingsText = document.createTextNode('The Books That I Read');
-    mainHeadings.appendChild(mainHeadingsText);
-
-    const ulTag2 = document.createElement('ul');
-    document.body.appendChild(ulTag2);
-
-    for (const keys of bookIds) {
-      const book = bookInfo[keys];
-      const liTag2 = document.createElement('li');
-      const attribute = document.createAttribute('id');
-      attribute.value = keys;
-      liTag2.setAttributeNode(attribute);
-      const headings = document.createElement('h2');
-      const headingsText = document.createTextNode(book.title);
-      headings.appendChild(headingsText);
-
-      const aboutLanguage = document.createElement('p');
-      const aboutLanguageSpan = document.createElement('span');
-      const aboutLanguageSpanText = document.createTextNode('Language: ');
-      aboutLanguageSpan.appendChild(aboutLanguageSpanText);
-      aboutLanguage.appendChild(aboutLanguageSpan);
-      const aboutLanguageText = document.createTextNode(book.language);
-      aboutLanguage.appendChild(aboutLanguageText);
+  function addNewTag(tagName, parent, options = {}) {
+    const elem = document.createElement(tagName);
+    parent.appendChild(elem);
+    for (const key of Object.keys(options)) {
+      const value = options[key];
+      if (key === 'text') {
+        elem.textContent = value;
+      } else {
+        elem.setAttribute(key, value);
+      }
+    }
+    return elem;
+  }
 
-      const aboutAuthor = document.createElement('p');
-      const aboutAuthorSpan = document.createElement('span');
-      const aboutAuthorSpanText = document.createTextNode('Author: ');
-      aboutAuthorSpan.appendChild(aboutAuthorSpanText);
-      aboutAuthor.appendChild(aboutAuthorSpan);
-      const aboutAuthorText = document.createTextNode(book.author);
-      aboutAuthor.appendChild(aboutAuthorText);
+  function addLabel(label, value, parent) {
+    const pTag = addNewTag('p', parent);
+    addNewTag('span', pTag, { text: `${label}: `, class: 'bold' });
+    addNewTag('span', pTag, { text: value });
+  }
 
-      liTag2.appendChild(headings);
-      liTag2.appendChild(aboutLanguage);
-      liTag2.appendChild(aboutAuthor);
-      ulTag2.appendChild(liTag2);
+  function createHtml() {
+    addNewTag('h1', document.body, { text: 'The Books That I Read' });
+    const ulTag2 = addNewTag('ul', document.body);
+    for (const bookId of bookIds) {
+      const book = bookInfo[bookId];
+      const liTag2 = addNewTag('li', ulTag2, { id: bookId });
+      addNewTag('h2', liTag2, { text: book.title });
+      addLabel('Language', book.language, liTag2);
+      addLabel('Author', book.author, liTag2);
     }
   }
 
   function addImage() {
     const keysOfBookCovers = Object.keys(bookCovers);
-
     for (const keys of keysOfBookCovers) {
```

Comment:
The `bookIds` array is what you need here, as you have done in line 129:

```js
for (const bookId of bookIds) {
```

In this way the `bookIds` array totally 'controls' which books are to be shown in the page and in which order. If you make changes to this array (remove items, reorder items) all else will follow along.

Reply:
Now I really got the logic of ordering the images ( before, here is a little dark in my mind but now it becomes clear)